roscala/src/main/scala/coop/rchain/rosette/prim/queue.scala

```diff
@@ -0,0 +1,196 @@
+import coop.rchain.rosette.prim.Prim._
+
+object queue {
+
+  object queueNew extends Prim {
+    override val name: String = "queue-new"
+    override val minArgs: Int = 0
+    override val maxArgs: Int = 0
+
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Queue] =
+      Right(Queue.create())
+  }
+
+  object queueDepth extends Prim {
+    override val name: String = "queue-depth"
+    override val minArgs: Int = 1
+    override val maxArgs: Int = 1
+
+    @checkTypeMismatch[Queue]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Fixnum] = {
+      val q = ctxt.argvec.elem.head.asInstanceOf[Queue]
+      Right(q.nElems.asInstanceOf[Fixnum])
+    }
+  }
+
+  object queueIsEmpty extends Prim {
+    override val name: String = "queue-empty?"
+    override val minArgs: Int = 1
+    override val maxArgs: Int = 1
+
+    @checkTypeMismatch[Queue]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, RblBool] = {
+      val q = ctxt.argvec.elem.head.asInstanceOf[Queue]
+      Right(RblBool(q.isEmpty()))
+    }
+  }
+
+  object queueEnqueue extends Prim {
+    override val name: String = "queue-enqueue"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Queue] =
+      ctxt.argvec.elem.head match {
+        case q: Queue =>
+          val ob = ctxt.argvec.elem(1)
+          Right(q.enqueue(ob))
+        case _ =>
+          Left(ArgumentMismatch("The first element should be a Queue"))
+      }
+  }
+
+  object queueDeQueue extends Prim {
+    override val name: String = "queue-dequeue"
+    override val minArgs: Int = 1
+    override val maxArgs: Int = 1
+
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Ob] =
+      ctxt.argvec.elem.head match {
+        case q: Queue =>
+          if (q.isEmpty()) {
+            Left(QueueEmptyError)
+          } else {
+            Right(q.dequeue().get)
+          }
+        case _ =>
+          Left(ArgumentMismatch("The first element should be a Queue"))
+      }
+  }
+
+  object queueRead extends Prim {
+    override val name: String = "queue-read"
+    override val minArgs: Int = 1
+    override val maxArgs: Int = 1
+
+    @checkTypeMismatch[Queue]
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Ob] = {
+      val q = ctxt.argvec.elem.head.asInstanceOf[Queue]
+      if (q.isEmpty()) {
+        Left(QueueEmptyError)
+      }
+      Right(q.head)
+    }
+  }
+
+  object queuePDequeue extends Prim {
+    override val name: String = "queue-pat-dequeue"
+    override val minArgs: Int = 2
+    override val maxArgs: Int = 2
+
+    @checkArgumentMismatch
+    override def fn(ctxt: Ctxt): Either[PrimError, Ob] =
+      (ctxt.argvec.elem.head, ctxt.argvec.elem(1)) match {
+        case (q: Queue, pat: Tuple) =>
+          if (q.isEmpty) {
```

Comment:
queue-pat-dequeue needs to return #absent for an empty queue, not PrimError.

Reply:
use ABSENT instead of a prim error

Reply:
`Left(Absent)`.
@kjmcqueen we want to keep the same errors as in Rosette but it's preferred to capture them the proper way.

Reply:
The queueRead object extends Prim trait, so the fn method must return Either[PrimError, Ob]. If we want return a Left result here, it must be a PrimError. Therefor, i add a new case object Absent which extends PrimError.

Reply:
Thanks.